frame/core-fellowship/src/lib.rs

```diff
@@ -0,0 +1,577 @@
+//!   anyone may call `bump` to remove the account's candidacy.
+//! - Pre-existing members may call `import` to have their rank recognised and be inducted into this
+//!   pallet (to gain a salary and allow for eventual promotion).
+//! - If, externally to this pallet, a member or candidate has their rank removed completely, then
+//!   `offboard` may be called to remove them entirely from this pallet.
+//!
+//! Note there is a difference between having a rank of 0 (whereby the account is a *candidate*) and
+//! having no rank at all (whereby we consider it *unranked*). An account can be demoted from rank
+//! 0 to become unranked. This process is called being offboarded and there is an extrinsic to do
+//! this explicitly when external factors to this pallet have caused the tracked account to become
+//! unranked. At rank 0, there is not a "demotion" period after which the account may be bumped to
+//! become offboarded but rather an "offboard timeout".
+//!
+//! Candidates may be introduced (i.e. an account to go from unranked to rank of 0) by an origin
+//! of a different privilege to that for promotion. This allows the possibility for even a single
+//! existing member to introduce a new candidate without payment.
+//!
+//! Only tracked/ranked accounts may submit evidence for their proof and promotion. Candidates
+//! cannot be approved - they must proceed only to promotion prior to the offboard timeout elapsing.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![recursion_limit = "128"]
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_arithmetic::traits::{Saturating, Zero};
+use sp_std::{marker::PhantomData, prelude::*};
+
+use frame_support::{
+	dispatch::DispatchResultWithPostInfo,
+	ensure,
+	traits::{tokens::Balance as BalanceTrait, EnsureOrigin, Get, RankedMembers},
+	BoundedVec, RuntimeDebug,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// The desired outcome for which evidence is presented.
+#[derive(Encode, Decode, Eq, PartialEq, Copy, Clone, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+pub enum Wish {
+	/// Member wishes only to retain their current rank.
+	Retention,
+	/// Member wishes to be promoted.
+	Promotion,
+}
+
+pub type Evidence<T, I> = BoundedVec<u8, <T as Config<I>>::EvidenceSize>;
+
+/// The status of the pallet instance.
+#[derive(Encode, Decode, Eq, PartialEq, Clone, TypeInfo, MaxEncodedLen, RuntimeDebug, Default)]
+pub struct ParamsType<Balance, BlockNumber> {
+	/// The amounts to be paid when a given grade is active.
+	active_salary: [Balance; 9],
+	/// The minimum percent discount when passive.
+	passive_salary: [Balance; 9],
+	/// The period between which unproven members become demoted.
+	demotion_period: [BlockNumber; 9],
+	/// The period between which members must wait before they may proceed to this rank.
+	min_promotion_period: [BlockNumber; 9],
```

Comment:
I'm guessing the 9 here denotes the number of ranks. Should we instead use a constant here, preferably from the fellowship pallet, to replace the hardcoded `9` here? Am generally in favour of self-documenting code.

Reply:
Can replace 9 with a constant if desired, however this is a pallet specifically for the core fellowship in which the nine ranks are quite fixed. I don't see much point in making it configurable at this stage.

Reply:
Ah, I meant a normal Rust const, not a const as a config item.